In [35]:
# Standard Imports
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
# Additional Imports
import os, json, math, time
import tmdbsimple as tmdb
from tqdm.notebook import tqdm_notebook

In [36]:
#Check for / create the data folder
import os
FOLDER = "Data/"
os.makedirs(FOLDER, exist_ok=True)
os.listdir(FOLDER)

['basics_clean.csv.gz',
 'final_tmdb_data_2000.csv.gz',
 'final_tmdb_data_2001.csv.gz',
 'title.akas.csv.gz',
 'title.ratings.csv.gz',
 'title_basics.csv.gz',
 'tmdb_api_results_2000.json',
 'tmdb_api_results_2001.json']

In [37]:
#Load TMDB API key
import json
with open('/Users/staichn/.secret/tmdb_api.json', 'r') as f:
    login = json.load(f)
## Display the keys of the loaded dict
login.keys()

dict_keys(['api-key'])

In [38]:
# Set API key to variable
tmdb.API_KEY =  login['api-key']

In [39]:
movie = tmdb.Movies(603)

info = movie.info()
info

{'adult': False,
 'backdrop_path': '/y9wuhlrqSHvhTLNVNwKMKe6HZzY.jpg',
 'belongs_to_collection': {'id': 2344,
  'name': 'The Matrix Collection',
  'poster_path': '/bV9qTVHTVf0gkW0j7p7M0ILD4pG.jpg',
  'backdrop_path': '/bRm2DEgUiYciDw3myHuYFInD7la.jpg'},
 'budget': 63000000,
 'genres': [{'id': 28, 'name': 'Action'},
  {'id': 878, 'name': 'Science Fiction'}],
 'homepage': 'http://www.warnerbros.com/matrix',
 'id': 603,
 'imdb_id': 'tt0133093',
 'original_language': 'en',
 'original_title': 'The Matrix',
 'overview': 'Set in the 22nd century, The Matrix tells the story of a computer hacker who joins a group of underground insurgents fighting the vast and powerful computers who now rule the earth.',
 'popularity': 59.281,
 'poster_path': '/f89U3ADr1oiB1s9GkdPOEpXUk5H.jpg',
 'production_companies': [{'id': 79,
   'logo_path': '/tpFpsqbleCzEE2p5EgvUq6ozfCA.png',
   'name': 'Village Roadshow Pictures',
   'origin_country': 'US'},
  {'id': 372,
   'logo_path': None,
   'name': 'Groucho II Film

In [40]:
# From Sample Solution function
def get_movie_with_rating(movie_id):
    ## Get movie and release dates
    movie = tmdb.Movies(movie_id)
    ## Construct output dict
    movie_info = movie.info()
    releases = movie.releases()
    for c in releases['countries']:
        if c['iso_3166_1'] == 'US':
            movie_info['certification'] = c['certification']
    return movie_info

In [41]:

# From Sample Solution function
def get_movie_with_rating(movie_id):
    ## Get movie and release dates
    movie = tmdb.Movies(movie_id)
    ## Construct output dict
    movie_info = movie.info()
    releases = movie.releases()
    for c in releases['countries']:
        if c['iso_3166_1'] == 'US':
            movie_info['certification'] = c['certification']
    return movie_info

In [42]:
get_movie_with_rating('tt0035423')

{'adult': False,
 'backdrop_path': '/ab5yL8zgRotrICzGbEl10z24N71.jpg',
 'belongs_to_collection': None,
 'budget': 48000000,
 'genres': [{'id': 35, 'name': 'Comedy'},
  {'id': 14, 'name': 'Fantasy'},
  {'id': 10749, 'name': 'Romance'}],
 'homepage': '',
 'id': 11232,
 'imdb_id': 'tt0035423',
 'original_language': 'en',
 'original_title': 'Kate & Leopold',
 'overview': "When her scientist ex-boyfriend discovers a portal to travel through time -- and brings back a 19th-century nobleman named Leopold to prove it -- a skeptical Kate reluctantly takes responsibility for showing Leopold the 21st century. The more time Kate spends with Leopold, the harder she falls for him. But if he doesn't return to his own time, his absence will forever alter history.",
 'popularity': 21.64,
 'poster_path': '/mUvikzKJJSg9khrVdxK8kg3TMHA.jpg',
 'production_companies': [{'id': 85,
   'logo_path': None,
   'name': 'Konrad Pictures',
   'origin_country': ''},
  {'id': 14,
   'logo_path': '/m6AHu84oZQxvq7n1rsvMN

In [43]:
YEARS_TO_GET = [2000,2001]

In [44]:
# Load in the dataframe from project part 1 as basics:
basics = pd.read_csv('data/basics_clean.csv.gz')

In [45]:
#Check that the dataframe loaded correctly
basics.head()

tconst titleType        primaryTitle       originalTitle  isAdult  \
0  tt0034413     short  Youth Gets a Break  Youth Gets a Break        0   
1  tt0035423     movie      Kate & Leopold      Kate & Leopold        0   
2  tt0040241     short     Color Rhapsodie     Color Rhapsodie        0   
3  tt0044326     short           Abstronic           Abstronic        0   
4  tt0044879     short             Mandala             Mandala        0   

   startYear  endYear  runtimeMinutes                  genres  
0       2001      NaN              20                   Short  
1       2001      NaN             118  Comedy,Fantasy,Romance  
2       2021      NaN               6                   Short  
3       2021      NaN               6                   Short  
4       2021      NaN               3                   Short

In [46]:
def write_json(new_data, filename): 
    """Adapted from: https://www.geeksforgeeks.org/append-to-json-file-using-python/"""    
    
    with open(filename,'r+') as file:
        # First we load existing data into a dict.
        file_data = json.load(file)
        ## Choose extend or append
        if (type(new_data) == list) & (type(file_data) == list):
            file_data.extend(new_data)
        else:
             file_data.append(new_data)
        # Sets file's current position at offset.
        file.seek(0)
        # convert back to json.
        json.dump(file_data, file)

In [47]:
# From Sample Solution function
def get_movie_with_rating(movie_id):
    ## Get movie and release dates
    movie = tmdb.Movies(movie_id)
    ## Construct output dict
    movie_info = movie.info()
    releases = movie.releases()
    for c in releases['countries']:
        if c['iso_3166_1'] == 'US':
            movie_info['certification'] = c['certification']
    return movie_info

In [48]:
# Start of OUTER loop
from tqdm.notebook import tqdm_notebook
for YEAR in tqdm_notebook(YEARS_TO_GET,desc='YEARS',position=0):
    #Defining the JSON file to store results for year
    JSON_FILE = f'{FOLDER}tmdb_api_results_{YEAR}.json'
    # Check if file exists
    file_exists = os.path.isfile(JSON_FILE)
    # If it does not exist: create it
    if file_exists == False:
    # save an empty dict with just "imdb_id" to the new json file.
        with open(JSON_FILE,'w') as f:
            json.dump([{'imdb_id':0}],f)
            #Saving new year as the current df


    df = basics.loc[ basics['startYear']==YEAR].copy()
    # saving movie ids to list
    movie_ids = df['tconst'].copy()#.to_list()
    previous_df = pd.read_json(JSON_FILE)
    movie_ids_to_get = movie_ids[~movie_ids.isin(previous_df['imdb_id'])]

    
    for movie_id in tqdm_notebook(movie_ids_to_get,
                                  desc=f'Movies from {YEAR}',
                                  position=1,
                                  leave=True):
        # Attempt to retrieve then data for the movie id
        try:
            temp = get_movie_with_rating(movie_id)  #This uses your pre-made function
            # Append/extend results to existing file using a pre-made function
            write_json(temp,JSON_FILE)
            # Short 20 ms sleep to prevent overwhelming server
            time.sleep(0.02)
            
        # If it fails,  make a dict with just the id and None for certification.
        except Exception as e:
            continue

YEARS:   0%|          | 0/2 [00:00<?, ?it/s]

Movies from 2000:   0%|          | 0/6046 [00:00<?, ?it/s]

KeyboardInterrupt: 

In [61]:
final_year_df = pd.read_json(JSON_FILE)
final_year_df.to_csv(f"{FOLDER}final_tmdb_data_{YEAR}.csv.gz", compression="gzip", index=False)

In [62]:
final_year_df.head()

imdb_id  adult                     backdrop_path belongs_to_collection  \
0          0    NaN                               NaN                   NaN   
1  tt0113026    0.0  /vMFs7nw6P0bIV1jDsQpxAieAVnH.jpg                  None   
2  tt0113092    0.0                              None                  None   
3  tt0116391    0.0                              None                  None   
4  tt0118694    0.0  /n4GJFGzsc7NinI1VeGDXIcQjtU2.jpg                  None   

       budget                                             genres  \
0         NaN                                                NaN   
1  10000000.0  [{'id': 35, 'name': 'Comedy'}, {'id': 10402, '...   
2         0.0           [{'id': 878, 'name': 'Science Fiction'}]   
3         0.0  [{'id': 18, 'name': 'Drama'}, {'id': 28, 'name...   
4    150000.0  [{'id': 18, 'name': 'Drama'}, {'id': 10749, 'n...   

                               homepage        id original_language  \
0                                   NaN       NaN               NaN   
1                                         62127.0                en   
2                                        110977.0                en   
3                                        442869.0                hi   
4  http://www.wkw-inthemoodforlove.com/     843.0                cn   

    original_title  ...     revenue  runtime  \
0              NaN  ...         NaN      NaN   
1  The Fantasticks  ...         0.0     86.0   
2    For the Cause  ...         0.0    100.0   
3             Gang  ...         0.0    152.0   
4             花樣年華  ...  12854953.0     99.0   

                                    spoken_languages    status  \
0                                                NaN       NaN   
1  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   
2  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   
3  [{'english_name': 'Hindi', 'iso_639_1': 'hi', ...  Released   
4  [{'english_name': 'Cantonese', 'iso_639_1': 'c...  Released   

                                             tagline                 title  \
0                                                NaN                   NaN   
1      Try to remember the first time magic happened       The Fantasticks   
2       The ultimate showdown on a forbidden planet.         For the Cause   
3                                                                     Gang   
4  Feel the heat, keep the feeling burning, let t...  In the Mood for Love   

   video  vote_average vote_count certification  
0    NaN           NaN        NaN           NaN  
1    0.0           5.5       22.0                
2    0.0           4.4        7.0                
3    0.0           0.0        0.0           NaN  
4    0.0           8.1     1891.0            PG  

[5 rows x 26 columns]

# How many movies had at least some valid financial information (values >0 for budget and revenue)?

In [63]:
budget_filter = final_year_df['budget'] > 0
revenue_filter = final_year_df['revenue'] > 0
financials = final_year_df.loc [budget_filter | revenue_filter, :]
financials.head()

imdb_id  adult                     backdrop_path  \
1   tt0113026    0.0  /vMFs7nw6P0bIV1jDsQpxAieAVnH.jpg   
4   tt0118694    0.0  /n4GJFGzsc7NinI1VeGDXIcQjtU2.jpg   
8   tt0119273    0.0  /fClJrAmJQ90zg9gWnfHVYSMiDTv.jpg   
12  tt0120263    0.0  /ynXVuylP8upazjz8lrqb1PEMkdR.jpg   
13  tt0120467    0.0  /knok3mNReKqPTplnnqz7E4dd7mD.jpg   

                                belongs_to_collection      budget  \
1                                                None  10000000.0   
4                                                None    150000.0   
8   {'id': 141086, 'name': 'Heavy Metal Collection...  15000000.0   
12                                               None         0.0   
13                                               None    120000.0   

                                               genres  \
1   [{'id': 35, 'name': 'Comedy'}, {'id': 10402, '...   
4   [{'id': 18, 'name': 'Drama'}, {'id': 10749, 'n...   
8   [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   
12  [{'id': 18, 'name': 'Drama'}, {'id': 35, 'name...   
13  [{'id': 18, 'name': 'Drama'}, {'id': 27, 'name...   

                                homepage       id original_language  \
1                                         62127.0                en   
4   http://www.wkw-inthemoodforlove.com/    843.0                cn   
8                                         16225.0                en   
12                                        34070.0                sv   
13                                        19085.0                en   

                original_title  ...     revenue  runtime  \
1              The Fantasticks  ...         0.0     86.0   
4                         花樣年華  ...  12854953.0     99.0   
8             Heavy Metal 2000  ...         0.0     88.0   
12  Sånger från andra våningen  ...     80334.0     98.0   
13                      Vulgar  ...     14904.0     87.0   

                                     spoken_languages    status  \
1   [{'english_name': 'English', 'iso_639_1': 'en'...  Released   
4   [{'english_name': 'Cantonese', 'iso_639_1': 'c...  Released   
8   [{'english_name': 'English', 'iso_639_1': 'en'...  Released   
12  [{'english_name': 'Russian', 'iso_639_1': 'ru'...  Released   
13  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   

                                              tagline  \
1       Try to remember the first time magic happened   
4   Feel the heat, keep the feeling burning, let t...   
8   not to survive the fight in the external world...   
12                                                      
13   Everyone loves a clown... some more than others.   

                          title  video  vote_average vote_count certification  
1               The Fantasticks    0.0           5.5       22.0                
4          In the Mood for Love    0.0           8.1     1891.0            PG  
8              Heavy Metal 2000    0.0           6.1      151.0             R  
12  Songs from the Second Floor    0.0           7.2      207.0                
13                       Vulgar    0.0           5.4       42.0             R  

[5 rows x 26 columns]

In [64]:
financials.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 254 entries, 1 to 892
Data columns (total 26 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   imdb_id                254 non-null    object 
 1   adult                  254 non-null    float64
 2   backdrop_path          222 non-null    object 
 3   belongs_to_collection  51 non-null     object 
 4   budget                 254 non-null    float64
 5   genres                 254 non-null    object 
 6   homepage               251 non-null    object 
 7   id                     254 non-null    float64
 8   original_language      254 non-null    object 
 9   original_title         254 non-null    object 
 10  overview               254 non-null    object 
 11  popularity             254 non-null    float64
 12  poster_path            250 non-null    object 
 13  production_companies   254 non-null    object 
 14  production_countries   254 non-null    object 
 15  releas

There are 254 movies with viable financial data.

# Q2:  How many movies are there in each of the certification categories (G/PG/PG-13/R)?

In [65]:
ratings_filter = financials['certification'].replace('', 'NR', regex = True)

In [66]:
financials['certification'].value_counts()

R        110
PG-13     57
          39
PG        18
G          8
NR         4
Name: certification, dtype: int64

# Q3: What is the average revenue per certification category?

In [70]:
movie_ratings = {'G', 'PG', 'PG-13', 'R', 'NR'}
rating_revenue = financials.groupby(ratings_filter, as_index=True)['revenue'].mean()
rating_revenue.head()

certification
G        1.053431e+08
NR       1.671715e+07
PG       7.597535e+07
PG-13    1.022932e+08
R        3.033181e+07
Name: revenue, dtype: float64

The mean film earnings per category are as follows:
G        $105,343,100
NR       $16,717,150
PG       $75,975,350
PG-13    $102,293,200
R        $30,331,810

# Q4: What is the average budget per certification category?

In [71]:
movie_ratings = {'G', 'PG', 'PG-13', 'R', 'NR'}
rating_revenue = financials.groupby(ratings_filter, as_index=True)['budget'].mean()
rating_revenue.head()

certification
G        4.500000e+07
NR       7.365019e+06
PG       4.566111e+07
PG-13    4.817368e+07
R        1.895812e+07
Name: budget, dtype: float64

The mean film budgets per category are as follows:
G        $45,000,000
NR       $7,365,019
PG       $45,661,110
PG-13    $48,173,680
R        $18,958,120